### DISCUSSION TITLE
Expectations memory/CPU usage

### DISCUSSION THREAD

**Discussion by JVApen** [pmd/pmd#3670] [Q&A] [answered]
On https://pmd.github.io/latest/pmd_userdocs_cpd.html it is mentioned that the application is memory hungry. On other pages it's indicated that it might take long to run this application.
What ain't clear from this is the actual impact.

If I have a large project, how will I know if I can run this application within reasonable time?
 Would it be possible to add some examples of (open source) projects on which this was ran and the time/memory used for it as well as the amount of files and MLoc of them?

**Comment by jsotuyod** [comment, marked as answer]
PMD ships 2 tools: PMD itself, and CPD. The CPU / memory needs for both are quite different.

## CPD

CPD in particular is a very memory hungry tool, as it has to tokenize all source code in order to find repeating patterns, so memory usage grows linearly with codebase size measured in tokens. The speed of the analysis can grow significantly, but not because of extraordinary CPU needs, but simply due to GC cycles as the memory needs increase (internally CPD caches stuff to avoid IOs which speed up analysis, but allows for GC reclamation).

Over time, CPD has gotten significantly better at memory management, to name a few:
* #919
* #198 (manually merged)
* #3385

So newer versions should not struggle even with large codebases such as Lucene (see https://github.com/pmd/pmd/issues/213). Moreover, newer JVMs, and using the newer / fancier GCs (ie: `-XX:+UseParallelGC` on Java 17) should produce additional gains even for the gnarliest and largest of codebases. However, I do not have readily available benchmarks for the current version.

## PMD

PMD on the other hand is not memory bound but IO / CPU bound. Files are analyzed one (per thread) at a time, so larger codebases solely need more analysis time, not necessarily more memory. Analysis speed has consistently improved over time, this is [a constant area of investment](https://github.com/pmd/pmd/pulls?q=is%3Apr+is%3Aclosed+label%3Afor%3Aperformance), and we have seen order-of-magnitude improvements since PMD 5.4.

Performance can vary greatly depending on your usage of the tool.
* Using incremental analysis allows to skip unmodified files between runs speeding up analysis once a baseline exists (see `-cache` flag)
* Using the auxclasspath (see `-auxclasspath` flag) produces the best possible results (in terms of false positives / negatives), but is costly in analysis time. PMD 7 will greatly improve this.
* Using multiple threads (ideally one per core) can greatly reduce analysis wall clock time (see `-t` flag)

PMD builds run regression tests on a bunch of large codebases routinely on all PRs. You can find reports such as [this one](https://chunk.io/pmd/abedd32e698d444fb3f62a6b09e61436/diff/index.html), where it takes ~15 minutes to analyze all 3 projects:
* Checkstyle 9.1
* OpenJDK 11
* Spring Framework 5.3.13 (all modules)

I do not have the exact SLoC count for these, but they are significant codebases. These analysis are not using cache, single threaded, without auxclasspath, running on commodity CI hardware provided by Github Actions.

Manually running PMD 6.41.0 against it's own codebase (all modules) which stands at ~376K SLoCs on my Macbook Pro 2020, 2.3 GHz Quad-Core Intel Core i7 with 4 threads, pruned cache and a full auxclasspath runs in 2:25. Not bad at all, and PMD 7 should be even faster.

As performance has so long been an investment area, I've long thought of including a page with charts showing speed comparisons between versions… but changes to the JVM versions and hardware makes it hard to maintain over time.

Adding no charts, but only some reference metrics could be an option, but again, would require regular updates with each release + your milage may vary, so I'm uncertain of their usefulness.

Overall, all PMD tools should be able to run smoothly on any project within reasonable times. Maybe we should simply drop that warning (which has been long standing), and let people report if they run into issues.

**Reply by jsotuyod** [reply]
I've just tried running the CPD 6.41.0 on the PMD codebase, it runs in 1.25s with a peak memory usage of 457MB for the whole JVM (default JVM memory settings and GC with Java 17). Turning on parallel GC reduces the memory usage to 322MB.

**Reply by JVApen** [reply]
Thanks for numbers. I'm convinced the code base I want to try it on is a magnitude larger than those indicated in the examples. However this already gives some confidence that it could be worth it.
When indicated it takes a lot of memory I was expecting something like 16GB for smaller projects, that doesn't seem to be the case.
